M2/Macaulay2/e/ExponentList.hpp

```diff
@@ -0,0 +1,229 @@
+/* Copyright 2006 by Michael E. Stillman */
+#pragma once
+
+#include <cstdio>  // for FILE
+#include <vector>  // for vector
+
+#include "engine-includes.hpp"  // for M2_arrayint, M2_ArrayString
+
+#include "ExponentVector.hpp"
+#include "buffer.hpp"
+
+class buffer;
+
+/** Implements operations on monomials represented as a list of (var, pow)'s
+ *
+ * This class implements monomial operations on lists of variable and exponent
+ * pairs. No allocation is done by any of these routines.
+ *
+ * The format for is an array [length, v1, e1, ..., vr, er]
+ * with v1 > v2 > ... > vr >= 0, and all exponents ei > 0 and length is 2r+1.
+ */
+// TODO: reimplement as a std::vector?
+template <class Exponent, bool legacy_length>
+class ExponentList;
+
+template <class Exponent, bool legacy_length>
+class ExponentListIterator;
+
+template <class E, bool L>
+class ExponentList
+{
+  using Iterator = ExponentListIterator<E, L>;
+  friend class ExponentListIterator<E, L>;
+
+ public:
+  typedef E Exponent;
+  typedef Exponent *Exponents;
+  typedef const Exponent *ConstExponents;
+  typedef typename std::make_unsigned<Exponent>::type HashExponent;
+  typedef gc_vector<Exponent> Vector;
+
+  static HashExponent computeHashValue(ConstExponents vp)
+  {
+    HashExponent hashval = *vp;
+    for (Iterator i = vp; i.valid(); ++i)
+      hashval = 4624296 * hashval + 2341 * i.var() + i.exponent();
+    return hashval;
+  }
+
+  static void elem_text_out(buffer &o, ConstExponents m, bool p_one = true)
```

Comment:
Yes, it was never used. The one in ExponentVector is also never used. I'd like to eventually combine these functions into one debug routine, because the only `elem_text_out` that actually gets used is in `monoid.cpp`.